<div style="border:solid green 3px; padding: 20px">

<b>Привет!</b>

Меня зовут Сороколетов Илья и я буду проверять твой проект. 

Предлагаю общаться на «**ты**», но если тебе будет комфортнее общаться на «**вы**», то сообщи об этом в комментариях)


Для твоего удобства, я буду выделять свои комментарии следующим образом:

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера v1:</b> Самые важные замечания. Они указывают на ключевые моменты, которые влияют на конечный результат проекта. </div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера v1:</b> Советы или замечания, которые помогут сделать твою работу лучше, но необязательны к выполнению.

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> Так я выделяю все остальные комментарии.</div>


Пометками <b>v1-v2-v3-...</b> я буду отмечать версию проверки. Так ты сможешь быстро найти мои новые комментарии.

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения, например вот так:
    
<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>
</div>

<b>Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии.</b> Если ты оставляешь свои комментарии, то делай это под моими, чтобы сообщения были расположены в хронологическом порядке. Всё это поможет выполнить повторную проверку твоего проекта быстрее. 


Перед отправкой работы я рекомендую нажимать Kernel -> Restart & Run All. Это перезапустит ядро и по очереди выполнит все ячейки. Так ты сможешь проверить, что всё работает корректно. Кнопка Kernel находится в панели сверху

Обязательно задавай вопросы если они возникнут, а я перехожу к проверке)

P.S. На всякий случай, я оставлю пустой шаблон для твоих комментариев ниже. Кликни два раза на мой комментарий, скопируй последние четыре строчки кода и вставляй их в пустую ячейку там, где ты хочешь оставить комментарий. Не забудь только перед этим сменить тип ячейки на Markdown. Быстро это можно сделать так: кликнуть на ячейку - нажать ESC - нажать M. 
    
    
<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Место для твоего комментария
</div>

# Исследование надежности заемщиков.

Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Описание данных
•	children — количество детей в семье
•	days_employed — общий трудовой стаж в днях
•	dob_years — возраст клиента в годах
•	education — уровень образования клиента
•	education_id — идентификатор уровня образования
•	family_status — семейное положение
•	family_status_id — идентификатор семейного положения
•	gender — пол клиента
•	income_type — тип занятости
•	debt — имел ли задолженность по возврату кредитов
•	total_income — ежемесячный доход
•	purpose — цель получения кредита

Для выполнения исследования в соответсвии с условиями Заказчика были сформулированы гипотезы:
1. Чем больше детей у кредитополучателя, тем реже возвращают кредит в срок.
2. "Семейные" кредитополучатели чаще возвращают кредит в срок, чем "не семейные".
3. Чем выше уровень дохода, тем чаще возвращают кредит в срок.
4. Цель кредита влияет на его возврат в срок.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Молодец, что добавляешь описание проекта. Это поможет тебе быстро вспомнить о чем проект, если спустя время ты решишь к нему вернуться)
</div>

Прежде, чем приступить к исследованию, необходимо выполнить обзор данных, проверить на ошибки и дубликаты.

### Шаг 1. Обзор данных

In [1]:
import pandas as pd

In [2]:
try:
    df = pd.read_csv(r'C:\Users\tanch\Desktop\data science\проекты\предобработка данных\data.csv') # чтение файла с данными
except:
    df = pd.read_csv('/datasets/data.csv')
df.head(30) # просмотр первых 30 строк таблицы

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0         женат / замужем                 0      F    сотрудник     0   
1         женат / замужем                 0      F    сотрудник     0   
2         женат / замужем                 0      M    сотрудник     0   
3         женат / замужем                 0      M    сотрудник     0   
4        гражданский брак                 1      F    пенсионер     0   
5        гражданский брак                 1      M    компаньон     0   
6         женат / замужем                 0      F    компаньон     0   
7         женат / замужем                 0      M    сотрудник     0   
8        гражданский брак                 1      F    сотрудник     0   
9         женат / замужем                 0      M    сотрудник     0   
10        женат / замужем                 0      M    компаньон     0   
11        женат / замужем                 0      F    сотрудник     0   
12       гражданский брак                 1      M    пенсионер     0   
13        женат / замужем                 0      F    сотрудник     0   
14       гражданский брак                 1      F    компаньон     1   
15        женат / замужем                 0      F    сотрудник     0   
16        женат / замужем                 0      F    сотрудник     0   
17       гражданский брак                 1      M    сотрудник     0   
18         вдовец / вдова                 2      F    пенсионер     0   
19              в разводе                 3      F    сотрудник     0   
20        женат / замужем                 0      M    сотрудник     0   
21       г

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера v1:</b> 
<br>Все библиотеки лучше импортировать отдельно в первой ячейке. Это позволит тебе в любой момент добавить ещё одну библиотеку без перезаписи переменных.
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Илья, спасибо за пояснения! Исправила!
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v2:</b>
<br>👍
</div>

При вызове первых 30 строк обнаружено:
1. пропуски NaN: строка 12, 26, 29 в столбцах 'days_employed' и 'total_income'
2. отрицательные значения в столбце 'days_employed'
3. "нереальные" цифры в столбце 'days_employed' (в описании к проекту 'days_employed' - трудовой стаж в днях): в 4 строке, если перевести в года, стаж составляет более 930 лет, что никак не может соответствовать действительности 
4. нет единого регистра написания в столбцах 'education' и 'family_status'.

Изучим подробнее данные в таблице.

In [3]:
df.info() # получение информации о данных в таблице

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов и 21525 строк.
Типы данных: пять столбцов с типом int64, два стоблца с типом float64, пять столбцов с типом object.

In [4]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Из явных ошибок в данных, в дополнение к выше обнаруженным:
1. значения min (-1) в столбце 'children', что в принципе не возможно
2. значение max (20) в столбце 'children', допустимо, но мало вероятно

Опеределим столбцы и количество строк с пропусками данных:

In [5]:
display(df.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Пропуски в данных есть только в столбцах 'days_employed' и 'total_income'.
Количество пропусков в обоих столбцах совпадает, возможно это пропуски в строках, которые относятся к одному клиенту.
Выясним так ли это:

In [6]:
 df[(df['days_employed'].isna() == True) & (df['total_income'].isna() == True)].info() # сравниваем пропуски в строках

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          2174 non-null   int64  
 1   days_employed     0 non-null      float64
 2   dob_years         2174 non-null   int64  
 3   education         2174 non-null   object 
 4   education_id      2174 non-null   int64  
 5   family_status     2174 non-null   object 
 6   family_status_id  2174 non-null   int64  
 7   gender            2174 non-null   object 
 8   income_type       2174 non-null   object 
 9   debt              2174 non-null   int64  
 10  total_income      0 non-null      float64
 11  purpose           2174 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


Если изучить две предыдущие таблицы, получается, что все пропуски находятся в двух столбцах и совпадают по строкам, что соответствуют одному клиенту.

Узнаем в процентах сколько всего пропусков от общего числа:

In [7]:
perс_skips = len(df[df['days_employed'].isna()]) / len(df) # расчет отношения пропусков к общему числу
print(f'Доля пропусков от общего числа строк {perс_skips:.2%}') # вывод результата в процентах с точностью 2 знака 

Доля пропусков от общего числа строк 10.10%


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Ещё можно посмотреть долю пропущенных значений среди всех данных следующим способом
    
    df.isna().mean()
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Обязательно запомню этот способ 
</div>

Возможные причины пропусков:
1. Клиент не предоставил полную информацию о себе
2. Сбой в системе
3. Человеческий фактор (опечатка, не верно введенная информация и т.д.)

Процент пропусков высок, поэтому удаление этих строк может повлиять на результат исследования. Необходимо заполнить пропуски.
Т.к. данные в столбцах 'days_employed' (общий трудовой стаж в днях) и 'total_income' (ежемесячный доход) относятся к количественным переменным, лучше заполнить пропуски медианным значением. В этом случае не будет большого влияния на общий результат.  

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Отличный анализ! Круто, что ты уже знаешь и используешь метод describe()
</div>

### Шаг 2.1 Заполнение пропусков

Заполним пропуски NaN в столбцах 'total_income'. Можно найти медиану по все столбцу, однако это будет не корректно, т.к. кредит берут разные группы по типам занятости. Поэтому используем группировку по типам занятости ('income_type') и найдем медианное значение для каждой группы по ежемесячному доходу ('total_income'). Таким образом, мы выведем медиану для каждой группы клиентов, что будет более корректно.

In [8]:
df.groupby('income_type')['total_income'].transform('median').head(15) #группировка по 'income_type', медиана по 'total_income'

0     142594.396847
1     142594.396847
2     142594.396847
3     142594.396847
4     118514.486412
5     172357.950966
6     172357.950966
7     142594.396847
8     142594.396847
9     142594.396847
10    172357.950966
11    142594.396847
12    118514.486412
13    142594.396847
14    172357.950966
Name: total_income, dtype: float64

Заполняем столбец 'total_income'. Проверяем таблицу.

In [9]:
df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform('median'))
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  118514.486412                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Пропуски в столбце 'total_income' заполнены медианными значениями в соответствии с типом занятости клиента.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>То что нужно! Это отличный способ для замены пропусков в данном случае.
</div>

### Шаг 2.2 Проверка данных на аномалии и исправления.

В данных столбца 'children' есть значения от -1 до 20. Определим варианты заполнения столбца:

In [10]:
df.groupby('children')['children'].count() # используем группировку по детям и подсчет для каждого варианта

children
-1        47
 0     14149
 1      4818
 2      2055
 3       330
 4        41
 5         9
 20       76
Name: children, dtype: int64

Итак: в таблице есть 47 строк, в которых указан -1 ребенок(не допустимое значение) и 76 строк со значение 20 детей (маловероятно). Варианты с количество детей от 3 до 5 вполне допустимы (многодетные семьи). Причина ошибки может быть как в человеке, так и в сбое при сборе данных. В любом случае, процент ошибки мал, что не должно сильно повлиять на результат дальнейшей обработки. Однако ее необходимо исправить: значение строк с -1 заменим на 0 детей, а строки со значением 20 заменим на 2. 

In [11]:
df.loc[df['children'] == -1, ['children']] = 0 # замена для строк со значением -1 на 0 детей
df.loc[df['children'] == 20, ['children']] = 2 # замена для строк со значением 20 на 2 ребенка
df.groupby('children')['children'].count() # проверка 

children
0    14196
1     4818
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Замена выполнена.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b>
<br>Отличное решение!
</div>

В столбце 'days_employed' обнаружены отрицательные значения и данные по трудовому стажу, которые не соответсвуют действительности. Т.к. по условию Заказчика нужно разобраться, влияет ли семейное положение 'family_status' и количество детей клиента 'children' на факт погашения кредита в срок, данные о трудовом стаже, скорее всего можно не учитывать. Поэтому приведем данные в столбце 'days_employed' к значению по модулю и заполним медианными значениями, которые будут соответствовать типу занятости.

In [12]:
df['days_employed'] = df['days_employed'].abs() # приводим к модулю
df['days_employed'] = df['days_employed'].fillna(df.groupby('income_type')['days_employed'].transform('median'))
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0  365213.306266         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  118514.486412                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Значения приведены по модулю числа. Пропуски в столбце 'days_employed' заполнены медианными значениями в соответсвии с типом занятости.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Супер) 

### Шаг 2.3. Изменение типов данных.

В столбце 'total_income' преобразуем данные из формата вещественных чисел в целочисленные. Используем метод astype()

In [13]:
df['total_income'] = df['total_income'].astype('int') # преобразование данных в целочисленный формат
df.head() # проверка таблицы

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Для удобства восприятия таблицы, столбце 'days_employed' также преобразуем данные из формата вещественных чисел в целочисленные:

In [14]:
df['days_employed'] = df['days_employed'].astype('int') # преобразование данных в целочисленный формат
df.head() # проверка таблицы

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Тут всё верно!
</div>

### Шаг 2.4. Удаление дубликатов.

Удаление дубликатов проведем в 2 этапа: выявим неявные, а затем явные дубликаты. Т.к. неявные дубликаты более скрыты, проверка на явные дубликаты будет разной до и после выявления. 

Проверка на явные дубликаты ДО выявления неявных дубликатов:

In [15]:
display(df.duplicated().sum()) # подсчет явных дубликатов

54

Проверим на неявные дубликаты столбец 'education':

In [16]:
df.groupby('education')['education'].count() # используем группировку по образованию и подсчет для каждого варианта

education
ВЫСШЕЕ                   274
Высшее                   268
НАЧАЛЬНОЕ                 17
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
Начальное                 15
Неоконченное высшее       47
СРЕДНЕЕ                  772
Среднее                  711
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
высшее                  4718
начальное                250
неоконченное высшее      668
среднее                13750
ученая степень             4
Name: education, dtype: int64

Обнаружены неявные дубликаты: данные записаны разным регистром (прописные и заглавные). Исправляем, приведя данные в нижний регистр: 

In [17]:
df['education'] = df['education'].str.lower() # преобразуем в нижний регистр 
df.groupby('education')['education'].count() #проверка таблицы

education
высшее                  5260
начальное                282
неоконченное высшее      744
среднее                15233
ученая степень             6
Name: education, dtype: int64

Проверим на неявные дубликаты столбец 'family_status':

In [18]:
df.groupby('family_status')['family_status'].count() # используем группировку по семейному статусу

family_status
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
гражданский брак          4177
женат / замужем          12380
Name: family_status, dtype: int64

Неявных дубликатов нет, однако есть разница в регистре. Исправляем, приведя данные в нижний регистр:

In [19]:
df['family_status'] = df['family_status'].str.lower()# преобразуем в нижний регистр
df.groupby('family_status')['family_status'].count() #проверка таблицы

family_status
в разводе                 1195
вдовец / вдова             960
гражданский брак          4177
женат / замужем          12380
не женат / не замужем     2813
Name: family_status, dtype: int64

Проверим на неявные дубликаты столбец 'income_type':

In [20]:
df.groupby('income_type')['income_type'].count() # группировка по типу занятости

income_type
безработный            2
в декрете              1
госслужащий         1459
компаньон           5085
пенсионер           3856
предприниматель        2
сотрудник          11119
студент                1
Name: income_type, dtype: int64

Неявных дубликатов нет. 

Проверим на неявные дубликаты столбец 'purpose':

In [21]:
df.groupby('purpose')['purpose'].count() # группировка по цели получения кредита

purpose
автомобили                                478
автомобиль                                495
высшее образование                        453
дополнительное образование                462
жилье                                     647
заняться высшим образованием              496
заняться образованием                     412
на покупку автомобиля                     472
на покупку подержанного автомобиля        479
на покупку своего автомобиля              505
на проведение свадьбы                     777
недвижимость                              634
образование                               447
операции с жильем                         653
операции с коммерческой недвижимостью     651
операции с недвижимостью                  676
операции со своей недвижимостью           630
покупка жилой недвижимости                607
покупка жилья                             647
покупка жилья для сдачи                   653
покупка жилья для семьи                   641
покупка коммерческой недви

Очень внушительный перечень, однако позиции в той или иной мере совпадают по смыслу. Думаю, что его можно сократить, объединив похожие позиции в одну смысловую группу. Таким образом, все действия с автомобилем (покупка, сделка и т.д.) можно свести в группу 'операции с автомобилем'. Аналогично поступить со всеми другими позициями (свадьба, образование, недвижимость). Для этого создадим функцию, которая вернет новые категории в столбец 'purpose_category' на основании данных из столбца 'purpose'  

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Супер! От дубликатов тоже избавились
</div>

### Шаг 2.5. Категоризация целей кредита.

In [22]:
def purpose_category(purpose):
    """
    Возвращает новую категорию по значению столбца 'purpose', используя правила:
    - 'операции с автомобилем', если в purpose есть подстрока 'автомобил';
    - 'операции с недвижимостью', если в purpose есть подстрока 'жиль' or 'недвижимост';
    - 'проведение свадьбы', если в purpose есть подстрока 'свадьб';
    - 'получение образование', если в purpose есть подстрока 'образован'.
    """
    
    if 'автомобил' in purpose:
        return 'операции с автомобилем'
    if 'жиль' in purpose or 'недвижимост' in purpose:
        return 'операции с недвижимостью'
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образован' in purpose:
        return 'получение образование'
    return 'уточните чель кредита' 

df['purpose_category'] = df['purpose'].apply(purpose_category)

display(df.head(15))

children  days_employed  dob_years            education  education_id  \
0          1           8437         42               высшее             0   
1          1           4024         36              среднее             1   
2          0           5623         33              среднее             1   
3          3           4124         32              среднее             1   
4          0         340266         53              среднее             1   
5          0            926         27               высшее             0   
6          0           2879         43               высшее             0   
7          0            152         50              среднее             1   
8          2           6929         35               высшее             0   
9          0           2188         41              среднее             1   
10         2           4171         36               высшее             0   
11         0            792         40              среднее             1   
12         0         365213         65              среднее             1   
13         0           1846         54  неоконченное высшее             2   
14         0           1844         56               высшее             0   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        118514   
13   женат / замужем                 0      F   сотрудник     0        130458   
14  гражданский брак                 1      F   компаньон     1        165127   

                              purpose          purpose_category  
0                       покупка жилья  операции с недвижимостью  
1             приобретение автомобиля    операции с автомобилем  
2                       покупка жилья  операции с недвижимостью  
3          дополнительное образование     получение образование  
4                     сыграть свадьбу        проведение свадьбы  
5                       покупка жилья  операции с недвижимостью  
6                   операции с жильем  операции с недвижимостью  
7                         образование     получение образование  
8               на проведение свадьбы        проведение свадьбы  
9             покупка жилья для семьи  операции с недвижимостью  
10               покупка недвижимости  операции с недвижимостью  
11  покупка коммерческой недвижимости  операции с недвижимостью  
12                    сыграть свадьбу        проведение свадьбы  
13            приобретение автомобиля    операции с автомобилем  
14         покупка жилой недвижимости  операции с недвижимостью

In [23]:
df.groupby('purpose_category')['purpose_category'].count()

purpose_category
операции с автомобилем       4315
операции с недвижимостью    10840
получение образование        4022
проведение свадьбы           2348
Name: purpose_category, dtype: int64

Проверка на явные дубликаты после выявления неявных дубликатов:

In [24]:
print(df.duplicated().sum()) # подсчет явных дубликатов

71


Очистим таблицу от явных дубликатов. Перезапишем индексы строк заново.

In [25]:
df = df.drop_duplicates().reset_index(drop=True) #удаляем явные дубликаты, индексы прописываются заново 
print(df.duplicated().sum()) # проверка 

0


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Основные цели кредита определены правильно! 👍
</div>


### Шаг 2.6. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

В таблице есть столбцы, которые по сути дублируют друг друга. Столбцы 'education' и 'education_id', а также столбцы 'family_status' и 'family_status_id'. Содержат одинаковую информацию, отличаются только форматом вывода. Поэтому лучше создать дополнительные датафреймы словарей, в которых будет сопоставляться id с его описанием.

Создадим датафрейм df_edu с данными из таблицы (столбцы 'education' и 'education_id'),  в котором будем сопоставлять идентификатор образования к его типу (id - '0' соответвует высшему образованию и т.д.)

In [26]:
df_edu = df[['education_id', 'education']] #создание дафафрейма по id и типу образования
display(df_edu.head(15))

education_id            education
0              0               высшее
1              1              среднее
2              1              среднее
3              1              среднее
4              1              среднее
5              0               высшее
6              0               высшее
7              1              среднее
8              0               высшее
9              1              среднее
10             0               высшее
11             1              среднее
12             1              среднее
13             2  неоконченное высшее
14             0               высшее

Почистим датафрейм от дубликатов. Проверим, чтобы для каждого id было свое описание:

In [27]:
df_edu = df_edu.drop_duplicates().reset_index(drop=True) #удаляем явные дубликаты, индексы прописываются заново
display(df_edu) # проверка 

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Повторим два предыдущих действия для столбцов 'family_status' и 'family_status_id'. Проверим результат.

In [28]:
df_family = df[['family_status_id', 'family_status']] #создание дафафрейма по id и по семейному положению
display(df_family.head(15))

family_status_id     family_status
0                  0   женат / замужем
1                  0   женат / замужем
2                  0   женат / замужем
3                  0   женат / замужем
4                  1  гражданский брак
5                  1  гражданский брак
6                  0   женат / замужем
7                  0   женат / замужем
8                  1  гражданский брак
9                  0   женат / замужем
10                 0   женат / замужем
11                 0   женат / замужем
12                 1  гражданский брак
13                 0   женат / замужем
14                 1  гражданский брак

In [29]:
df_family = df_family.drop_duplicates().reset_index(drop=True) #удаляем явные дубликаты, индексы прописываются заново
display(df_family) # проверка 

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

Удалим из исходного датафрейма столбцы 'education' и 'family_status', оставив только их идентификаторы: 'education_id' и 'family_status_id'.

In [30]:
df = df.drop(['education', 'family_status'], axis=1) # удаление столбцов 'education' и 'family_status'
display(df.head(15)) # проверка

children  days_employed  dob_years  education_id  family_status_id gender  \
0          1           8437         42             0                 0      F   
1          1           4024         36             1                 0      F   
2          0           5623         33             1                 0      M   
3          3           4124         32             1                 0      M   
4          0         340266         53             1                 1      F   
5          0            926         27             0                 1      M   
6          0           2879         43             0                 0      F   
7          0            152         50             1                 0      M   
8          2           6929         35             0                 1      F   
9          0           2188         41             1                 0      M   
10         2           4171         36             0                 0      M   
11         0            792         40             1                 0      F   
12         0         365213         65             1                 1      M   
13         0           1846         54             2                 0      F   
14         0           1844         56             0                 1      F   

   income_type  debt  total_income                            purpose  \
0    сотрудник     0        253875                      покупка жилья   
1    сотрудник     0        112080            приобретение автомобиля   
2    сотрудник     0        145885                      покупка жилья   
3    сотрудник     0        267628         дополнительное образование   
4    пенсионер     0        158616                    сыграть свадьбу   
5    компаньон     0        255763                      покупка жилья   
6    компаньон     0        240525                  операции с жильем   
7    сотрудник     0        135823                        образование   
8    сотрудник     0         95856              на проведение свадьбы   
9    сотрудник     0        144425            покупка жилья для семьи   
10   компаньон     0        113943               покупка недвижимости   
11   сотрудник     0         77069  покупка коммерческой недвижимости   
12   пенсионер     0        118514                    сыграть свадьбу   
13   сотрудник     0        130458            приобретение автомобиля   
14   компаньон     1        165127         покупка жилой недвижимости   

            purpose_category  
0   операции с недвижимостью  
1     операции с автомобилем  
2   операции с недвижимостью  
3      получение образование  
4         проведение свадьбы  
5   операции с недвижимостью  
6   операции с недвижимостью  
7      получение образование  
8         проведение свадьбы  
9   операции с недвижимостью  
10  операции с недвижимостью  
11  операции с недвижимостью  
12        проведение свадьбы  
13    операции с автомобилем  
14  операции с недвижимостью

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Тут всё правильно 👌
</div>

### Шаг 2.7. Категоризация дохода.

Дополним таблицу столбцом 'total_income_category', в котором будут указаны категории кредитополучателей в соотвествии с доходом. Для этого создадим функцию:

In [31]:
def total_income_category(total_income):
    """
    Возвращает новую категорию по значению столбца 'total_income', используя правила:
    - если доход от 0 до 30000, то категория Е;
    - если доход от 30001 до 50000, то категория D;
    - если доход от 50001 до 200000, то категория C;
    - если доход от 200001 до 1000000, то категория B;
    - если доход от 1000001 и выше, то категория А.
    """
    
    if 0 <= total_income <= 30000:
        return 'E'
    if 30001 <= total_income <= 50000:
        return 'D'
    if 50001 <= total_income <= 200000:
        return 'C'
    if 200001 <= total_income <= 1000000:
        return 'B'
    if 1000001 <= total_income:
        return 'A'
    return 'уточните ежемесячный доход' 

df['total_income_category'] = df['total_income'].apply(total_income_category)

display(df.head(15))

children  days_employed  dob_years  education_id  family_status_id gender  \
0          1           8437         42             0                 0      F   
1          1           4024         36             1                 0      F   
2          0           5623         33             1                 0      M   
3          3           4124         32             1                 0      M   
4          0         340266         53             1                 1      F   
5          0            926         27             0                 1      M   
6          0           2879         43             0                 0      F   
7          0            152         50             1                 0      M   
8          2           6929         35             0                 1      F   
9          0           2188         41             1                 0      M   
10         2           4171         36             0                 0      M   
11         0            792         40             1                 0      F   
12         0         365213         65             1                 1      M   
13         0           1846         54             2                 0      F   
14         0           1844         56             0                 1      F   

   income_type  debt  total_income                            purpose  \
0    сотрудник     0        253875                      покупка жилья   
1    сотрудник     0        112080            приобретение автомобиля   
2    сотрудник     0        145885                      покупка жилья   
3    сотрудник     0        267628         дополнительное образование   
4    пенсионер     0        158616                    сыграть свадьбу   
5    компаньон     0        255763                      покупка жилья   
6    компаньон     0        240525                  операции с жильем   
7    сотрудник     0        135823                        образование   
8    сотрудник     0         95856              на проведение свадьбы   
9    сотрудник     0        144425            покупка жилья для семьи   
10   компаньон     0        113943               покупка недвижимости   
11   сотрудник     0         77069  покупка коммерческой недвижимости   
12   пенсионер     0        118514                    сыграть свадьбу   
13   сотрудник     0        130458            приобретение автомобиля   
14   компаньон     1        165127         покупка жилой недвижимости   

            purpose_category total_income_category  
0   операции с недвижимостью                     B  
1     операции с автомобилем                     C  
2   операции с недвижимостью                     C  
3      получение образование                     B  
4         проведение свадьбы                     C  
5   операции с недвижимостью                     B  
6   операции с недвижимостью                     B  
7      получение образование                     C  
8         проведение свадьбы                     C  
9   операции с недвижимостью                     C  
10  операции с недвижимостью                     C  
11  операции с недвижимостью                     C  
12        проведение свадьбы                     C  
13    операции с автомобилем                     C  
14  операции с недвижимостью                     C

In [32]:
df.groupby('total_income_category')['total_income_category'].count() # проверка функции

total_income_category
A       25
B     5042
C    16015
D      350
E       22
Name: total_income_category, dtype: int64

Функция отработала верно.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Категоризация клиентов по уровню дохода произведена верно
</div>

### 2.8. Проверка гипотез

Проверка данных на корректность выполнена. Для выполнения исследования в соответсвии с условиями Заказчика необходимо проверить гипотезы:
1. Чем больше детей у кредитополучателя, тем реже возвращают кредит в срок.
2. "Семейные" кредитополучатели чаще возвращают кредит в срок, чем "не семейные".
3. Чем выше уровень дохода, тем чаще возвращают кредит в срок.
4. Цель кредита влияет на его возврат в срок.

### Гипотеза №1

Проверим первую гипотезу: "Чем больше детей у кредитополучателя, тем реже возвращают кредит в срок." 
Создадим сводную таблицу, в которой будет отображаться группировка по количеству детей ('children') и суммарная задолженность по погащению кредита в срок ('debt'). Найдем процентное соотнощение ('ratio %') задолженности по погащению кредита в срок к общему числу кредиполучателей в каждой группе ('children').

In [33]:
children_pivot = df.pivot_table(index=['children'], values='debt', aggfunc= ('count', 'sum')) #сводная таблица по 'children' и 'debt' 
children_pivot['ratio, %'] = (children_pivot['sum'] / children_pivot['count']) * 100 # соотношение в процентах
display(children_pivot.head(10))

count   sum  ratio, %
children                       
0         14138  1064  7.525817
1          4808   444  9.234609
2          2128   202  9.492481
3           330    27  8.181818
4            41     4  9.756098
5             9     0  0.000000

Вывод по таблице: в целом процент не благонадежных кредиполучателей около 10% в каждой категории (кроме тех, у кого 5 детей - там 0%, что возможно связано с редкостью такой группы). Таким образом, рассмотрим только данные где есть результат. Исходя из этого видно, что в группе где нет детей, процент все-таки ниже, чем в группах с детьми. Причем, в "детских" группах процент задолженности выплаты по кредиту колеблется. Но динамика роста есть (если исключить группу с 3-мя детьми). Что касается группы с 3-мя детьми, возможно не хватает данных для более точного результата.

Вывод по гипотезе: предположение подтвердилось. Кредитополучатели с детьми склонны к просрочке по выплате кредита.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Верно! Отлично, что ты используешь сводную таблицу
</div>

### Гипотеза №2

Гипотеза сформулированна следующим образом: "Семейные" кредитополучатели чаще возвращают кредит в срок, чем "не семейные". Для проверки гипотезы создадим сводную таблицу, в которой будет отображаться группировка по идентификатору семейного положения ('family_status_id') и суммарная задолженность по погащению кредита в срок ('debt'). Найдем процентное соотнощение ('ratio %') задолженности по погащению кредита в срок к общему числу кредиполучателей в каждой группе ('family_status_id'). Для удобства восприятия внесем дополнительную информацию из датафрема df_family по столбцу 'family_status_id' с расшифровкой каждого id.

In [34]:
famile_pivot = pd.pivot_table(df, index=['family_status_id'], values='debt', aggfunc= ('count', 'sum'))
famile_pivot['ratio, %'] = (famile_pivot['sum'] / famile_pivot['count']) * 100
famile_status_pivot = df_family.merge(famile_pivot, on='family_status_id') # объединяем обе таблицы
display(famile_status_pivot.head(10))

family_status_id          family_status  count  sum  ratio, %
0                 0        женат / замужем  12339  931  7.545182
1                 1       гражданский брак   4151  388  9.347145
2                 2         вдовец / вдова    959   63  6.569343
3                 3              в разводе   1195   85  7.112971
4                 4  не женат / не замужем   2810  274  9.750890

Вывод по таблице: как и в предыдущей таблице в целом процент не благонадежных кредиполучателей около 10% в каждой категории. Однако разрыв между категорией между официальными отношениями (женат/замужем) и гражданским браком есть. И не в пользу гражданского брака. Так же, склонны нарушать обязательства по выплате и те, кто относятся к группе не женат/не замужем. 

Вывод по гипотезе: она подтвердилась. Кредитополучатели, которые состоят в офицальном браке чаще вовремя выплачивают кредит, чем те, кто состоит в гражданском браке, или не имеет партнера. 

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Тут всё правильно
</div>

### Гипотеза №3

Формулировка гипотезы: Чем выше уровень дохода, тем чаще возвращают кредит в срок. Для проверки гипотезы создадим сводную таблицу, в которой будет отображаться группировка по уровню дохода ('total_income_category') и суммарная задолженность по погащению кредита в срок ('debt'). Найдем процентное соотнощение ('ratio %') задолженности по погащению кредита в срок к общему числу кредиполучателей в каждой группе ('total_income_category').

В основной таблице нет пояснения для категорий по уровню дохода. Создадим дополнительную таблицу, в которой для каждой катерии доходов будут указаны диапазоны значений и включим ее в сводную таблицу по столбцу 'total_income_category'

In [35]:
income_interval = [
    ['A','доход от 1000001 и выше'],
    ['B','доход от 200001 до 1000000'],
    ['C','доход от 50001 до 200000'],
    ['D', 'доход от 30001 до 50000'],
    ['E', 'доход от 0 до 30000'],
]
income_value = ['total_income_category', 'income_interval'] # создаём таблицу

income_descript = pd.DataFrame(data=income_interval, columns=income_value) # создаём таблицу и сохраняем её в переменную income_descript
display(income_descript) # проверка

total_income_category             income_interval
0                     A     доход от 1000001 и выше
1                     B  доход от 200001 до 1000000
2                     C    доход от 50001 до 200000
3                     D     доход от 30001 до 50000
4                     E         доход от 0 до 30000

In [36]:
total_income_pivot = df.pivot_table(index=['total_income_category'], values='debt', aggfunc= ('count', 'sum')) #сводная таблица по 'children' и 'debt' 
total_income_pivot['ratio, %'] = (total_income_pivot['sum'] / total_income_pivot['count']) * 100 # соотношение в процентах
income_status_pivot = income_descript.merge(total_income_pivot, on='total_income_category') # объединяем обе таблицы
display(income_status_pivot.head(10))

total_income_category             income_interval  count   sum  ratio, %
0                     A     доход от 1000001 и выше     25     2  8.000000
1                     B  доход от 200001 до 1000000   5042   356  7.060690
2                     C    доход от 50001 до 200000  16015  1360  8.492039
3                     D     доход от 30001 до 50000    350    21  6.000000
4                     E         доход от 0 до 30000     22     2  9.090909

Вывод по таблице: порог не благонадежных кредиполучателей около 10% в каждой категории сохраняется. Самые злостные нарушители по срокам погашения кредита относятся к категории 'E', т.е. с самым минимальным уровнем дохода. При этом категория 'D', показывает самый низкий процент просрочки - 6%. Группы с высоким уровнем дохода ('А - C') так же как и категория 'E', склонны к задержкам погашения кредита в срок.

Вывод по гипотезе: нельзя однозначно утверждать, что есть прямая зависимость погащения кредита в срок от уровня дохода. Однако кредитополучатели с уровнем дохода категории 'D' предпочтительнее. 

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Согласен! Идём дальше)
</div>

### Гипотеза №4

Данная гипотеза предполагает, что цель получения кредита влияет на срок погашения. Т.е. кредит связанный с операциями по недвижимости или с операциями по автомобилям (как более финансозатратные) будут закрываться чаще без просрочек по кредиту. 
Для проверки гипотезы создадим сводную таблицу, в которой будет отображаться группировка по целям кредита  ('purpose_category') и суммарная задолженность по погащению кредита в срок ('debt'). Найдем процентное соотнощение ('ratio %') задолженности по погащению кредита в срок к общему числу кредитных целей в каждой группе ('purpose_category').

In [37]:
purpose_pivot = df.pivot_table(index=['purpose_category'], values='debt', aggfunc= ('count', 'sum')) #сводная таблица по 'children' и 'debt' 
purpose_pivot['ratio, %'] = (purpose_pivot['sum'] / purpose_pivot['count']) * 100 # соотношение в процентах
display(purpose_pivot.head(10))

count  sum  ratio, %
purpose_category                              
операции с автомобилем     4306  403  9.359034
операции с недвижимостью  10811  782  7.233373
получение образование      4013  370  9.220035
проведение свадьбы         2324  186  8.003442

Выводы по таблице: кредиты связанные с операциями по недвижимостью меньше подвержены просрочками по возврату в срок. А вот кредиты, цель которых была связана с автомобилем, наоборот, имеет самый высокий процент по не возврату кредита в срок.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b>
<br>Абсолютно верно!
</div>

### Общий вывод

В исследовании было сформулировано 4 гипотезы:
1. Чем больше детей у кредитополучателя, тем реже возвращают кредит в срок.
2. "Семейные" кредитополучатели чаще возвращают кредит в срок, чем "не семейные".
3. Чем выше уровень дохода, тем чаще возвращают кредит в срок.
4. Цель кредита влияет на его возврат в срок.

Все рассмотренные категории показали, что не более 10% кредитополучателей склонны к задержкам по выплатам по кредиту. <br> 
Самые выгодные кредитополучатели это семейные пары, которые состоят в законном браке и не имеют детей. При этом, средний уровень дохода семьи должен быть от 30000 до 50000 и цель кредита которых направлена на операции с недвижимостью. <br>
Самые неблагонадежные категории  - это не имеющие официальных отношений, либо с низким уровнем дохода. При этом цель кредита получение быстрого результата в краткосрочной перспективе.
Нельзя утверждать, что все гипотезы были подтверждены или опровергнуты. <br> 
В каждом результате есть "исключения". <br>
В первой гипотезе недостаточно данных для многодетныйх семей (0% просрочке по кредиту в группе, где 5 детей).<br>
Во второй гипотезе есть две категории (вдовец/вдова (6,56%), в разводе (7,11%)), которые вызывают сомнения. Там самый низкий показатель по срокам погащения кредита в срок. В официальном понятии такие категории допустимы, однако эти же категории не исключают "гражданского брака". <br>
В третьей гипотезе явно не наблюдается какой-то закономерности. Изменение задолженности по выплате кредита  наблюдается скачками: рост (9,09%), снижение (6,00%), рост (8,49%), снижение (7,06%), рост (8,00%). <br>
Четвертая гипотеза тоже не дала однозначного результата. Возможно надо расширить параметры, которые могут более полно исследовать, как влияет цель кредита на его возврат в срок.<br>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Общий вывод верный 👍
</div>

<div class="alert alert-success">
<b>✔️ Итоговый комментарий ревьюера v1:</b>
<br>Я бы мог принять твою работу с первого раза, но хочу дать тебе возможность задать мне вопросы) Если вопросов нет, то я с радостью приму проект)
</div>


<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Илья спасибо за проверку! Я еще новичок в написании таких проектов. Вопрос такой: возможны ли сбои в работе Jupiter? Поясню: когда я написала код функции, при ее запуске выдавали ошибки (синаксис в основном). Вообщем, я честно пыталась все исправить, но каждый раз ошибка возникала снова. Логика функции была верная, я бесконечно проверяла код. Ошибка! Короче, я сохранила последний код, который выдавал ошибку. И закрыла блокнот, решив на следующий день его опять исправить. Но когда я зашла в следующий раз и запустила код - все заработало! Я так и не поняла, почему раннее код с ошибкой запустился! Как такое может быть? 
<br> Может есть какой-то лайфхак при работе с такими проектами? Еще раз благодарю!
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v2:</b>
<br>Очень сложно ответить на твой вопрос, не видя ошибки) Бывают проблемы с юпитером, но они обычно связаны с его общий состоянием, про такой случай слышу в первые. Возможно, дело было в том, что на следующий день ты последовательно воспроизвела все ячейки и у тебя всё заработало) Лайфхак - сделать скриншот и отнести в чат с преподавателем. Он сможет подсказать)
</div>

<div class="alert alert-success">
<b>✔️ Итоговый комментарий ревьюера v2:</b>
<br>Рад был поработать над проверкой твоего проекта! Желаю удачи в дальнейшем обучении! 😉
</div>